### PR TITLE
CC-12684: Minimal avalara integration.

### PR DESCRIPTION
**Pull request by AsonUnique**
- Developer(s): @AsonUnique

- Ticket: https://spryker.atlassian.net/browse/CC-12864

- Academy PR: -

- Release Group: https://release.spryker.com/release-groups/view/3457

- merge: merge

#### Release Table

   Module                | Release Type         | Constraint Updates         |
   :--------------------- | :------------------------ | :--------------------- |
   AvalaraTax            | minor (currently 0.0.x)  |                        |

-----------------------------------------

##### Change log

Initial Release

- The module allows the `Spryker` project to connect to the `Avalara` tax service.

### CODE REVIEW

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+          composer --version
+          composer install
+      - name: Setup PHP
+        uses: shivammathur/setup-php@v2
+        with:
+          php-version: ${{ matrix.php }}
+          extensions: mbstring, intl, pdo_mysql
+          tools: composer:v2
+
+#       suite-tests:
```
Comment (abitskil) [inline .github/workflows/ci.yml]: * Remove commented part.

File: composer.json [hunk 2]
```diff
@@ -0,0 +1,59 @@
+{
+    "name": "spryker-eco/avalara-tax",
+    "type": "library",
+    "description": "AvalaraTax module",
+    "license": "MIT",
+    "require": {
+        "avalara/avataxclient": "^21.1.1",
+        "php": ">=7.3",
+        "spryker-eco/avalara-tax-extension": "1.0.0",
```
Comment (abitskil) [inline composer.json]: * `0.1.0` - beta module.

File: composer.json [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+        "spryker/calculation-extension": "^1.0.0",
+        "spryker/cart-extension": "^4.0.0",
```
Comment (abitskil) [inline composer.json]: * `^4.2.0` => `ItemExpanderPluginInterface`.

File: composer.json [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+        "spryker/checkout-extension": "^1.2.0",
+        "spryker/checkout-rest-api-extension": "^1.0.0",
```
Comment (abitskil) [inline composer.json]: * `^1.5.0` => `ReadCheckoutDataValidatorPluginInterface`.

File: src/SprykerEco/Zed/AvalaraTax/Business/Executor/AvalaraTransactionExecutor.php [hunk 5]
```diff
@@ -0,0 +1,163 @@
...
+            );
+
+        $transactionBuilder = $this->avalaraTransactionBuilder->buildCreateTransaction(
+            $avalaraCreateTransactionRequestTransfer,
+            $transactionTypeId,
+        );
+
+        $avalaraApiLogTransfer = (new AvalaraApiLogTransfer())
+            ->setRequest($this->utilEncodingService->encodeJson($transactionBuilder->toArray()))
+            ->setTransactionType($transactionTypeId);
```
Comment (abitskil) [inline src/SprykerEco/Zed/AvalaraTax/Business/Executor/AvalaraTransactionExecutor.php:106]: * Use store name from a `CalculableObjectTransfer`.
* Remove the dependecy to a `Store` module.

File: src/SprykerEco/Zed/AvalaraTax/AvalaraTaxConfig.php [hunk 6]
```diff
@@ -0,0 +1,143 @@
...
+    }
+
+    /**
+     * @api
+     *
+     * @return string
+     */
+    public function getDefaultCustomerCode(): string
+    {
+        return 'TESTCUSTOMER';
```
Comment (abitskil) [inline src/SprykerEco/Zed/AvalaraTax/AvalaraTaxConfig.php:102]: * Can be empty ?

File: src/SprykerEco/Zed/AvalaraTax/AvalaraTaxConfig.php [hunk 7]
```diff
@@ -0,0 +1,143 @@
...
+
+use Spryker\Zed\Kernel\AbstractBundleConfig;
+use SprykerEco\Shared\AvalaraTax\AvalaraTaxConstants;
+
+class AvalaraTaxConfig extends AbstractBundleConfig
+{
+    /**
+     * @uses \Avalara\DocumentType::C_SALESORDER
+     */
+    public const AVALARA_TRANSACTION_TYPE_ID_SALES_ORDER = 0;
```
Comment (abitskil) [inline src/SprykerEco/Zed/AvalaraTax/AvalaraTaxConfig.php]: * If constants are not used outside the class, mark them as "protected".

Comment (abitskil) [reply src/SprykerEco/Zed/AvalaraTax/AvalaraTaxConfig.php]: Generate a `psalm-report.json` for a module.

File: src/SprykerEco/Zed/AvalaraTax/AvalaraTaxConfig.php (hunk 6 above)
Comment (AsonUnique) [reply src/SprykerEco/Zed/AvalaraTax/AvalaraTaxConfig.php:102]: No, I've tried to send an empty string as a customer code, but Avalara responded with an error.

File: .github/workflows/ci.yml (hunk 1 above)
Comment (abitskil) [reply .github/workflows/ci.yml]: Resolved.

File: composer.json (hunk 2 above)
Comment (abitskil) [reply composer.json]: Resolved.

File: composer.json (hunk 3 above)
Comment (abitskil) [reply composer.json]: Resolved.

File: composer.json (hunk 4 above)
Comment (abitskil) [reply composer.json]: Resolved.

File: src/SprykerEco/Zed/AvalaraTax/Business/Executor/AvalaraTransactionExecutor.php (hunk 5 above)
Comment (abitskil) [reply src/SprykerEco/Zed/AvalaraTax/Business/Executor/AvalaraTransactionExecutor.php:106]: Resolved.

File: src/SprykerEco/Zed/AvalaraTax/AvalaraTaxConfig.php (hunk 6 above)
Comment (abitskil) [reply src/SprykerEco/Zed/AvalaraTax/AvalaraTaxConfig.php:102]: Resolved.

File: src/SprykerEco/Zed/AvalaraTax/Business/Builder/AvalaraTransactionBuilder.php [hunk 8]
```diff
@@ -0,0 +1,155 @@
...
+        if ($avalaraLineItemTransfer->getTaxIncluded()) {
+            $transactionBuilder->withLineTaxIncluded();
+        }
+
+        if ($avalaraLineItemTransfer->getShippingAddress() !== null) {
+            $transactionBuilder = $this->addItemLevelAddress($transactionBuilder, $avalaraLineItemTransfer->getShippingAddressOrFail());
+        }
+
+        if ($avalaraLineItemTransfer->getSourceAddress() !== null) {
+            $transactionBuilder = $this->addItemLevelAddress($transactionBuilder, $avalaraLineItemTransfer->getSourceAddress());
```
Comment (abitskil) [inline src/SprykerEco/Zed/AvalaraTax/Business/Builder/AvalaraTransactionBuilder.php]: * orFail().

File: src/SprykerEco/Zed/AvalaraTax/AvalaraTaxConfig.php (hunk 7 above)
Comment (abitskil) [reply src/SprykerEco/Zed/AvalaraTax/AvalaraTaxConfig.php]: Resolved.

File: src/SprykerEco/Zed/AvalaraTax/Business/Builder/AvalaraTransactionBuilder.php (hunk 8 above)
Comment (abitskil) [reply src/SprykerEco/Zed/AvalaraTax/Business/Builder/AvalaraTransactionBuilder.php]: Resolved.

**Review by abitskil — APPROVED** [review]

**Pull request merged**